cynthion/python/src/gateware/analyzer/fifo.py

```diff
@@ -19,16 +19,31 @@ def __init__(self):
     def elaborate(self, platform):
         m = Module()
 
-        cycle_count = Signal()
-        m.d.sync += cycle_count.eq(cycle_count + 1)
+        # To ensure at least 2 cycles of stable data for clock domain transfer,
+        # use an intermediate buffer with a cycle counter.
+        buffer  = StreamInterface(payload_width=16)
+        counter = Signal(range(2))
 
         with m.If(~self.output.valid | self.output.ready):
             m.d.usb += [
-                self.output.payload .eq(self.input.payload),
-                self.output.valid   .eq(self.input.valid),
-                self.output.last    .eq(self.input.last),
+                self.output.payload .eq(buffer.payload),
+                self.output.valid   .eq(buffer.valid),
+                self.output.last    .eq(buffer.last)
             ]
-            m.d.comb += self.input.ready.eq(cycle_count)
+            with m.If(counter != 0):
+                m.d.sync += buffer.valid.eq(0)
+
+        m.d.sync += counter.eq(Mux(counter == 1, counter, counter + 1))
+
+        # Fill the intermediate buffer and restart the counter.
+        with m.If(~buffer.valid):
+            m.d.sync += [
+                buffer.payload      .eq(self.input.payload),
+                buffer.valid        .eq(self.input.valid),
+                buffer.last         .eq(self.input.last),
+                counter             .eq(0),
+            ]
+            m.d.comb += self.input.ready.eq(1)
```

Comment:
`counter` is only ever 0 or 1, so it can just be a boolean flag as far as I can see. I tried this and it worked:

```python
        # To ensure at least 2 cycles of stable data for clock domain transfer,
        # use an intermediate buffer with a ready flag.
        buffer  = StreamInterface(payload_width=16)
        ready   = Signal()

        with m.If(~self.output.valid | self.output.ready):
            m.d.usb += [
                self.output.payload .eq(buffer.payload),
                self.output.valid   .eq(buffer.valid),
                self.output.last    .eq(buffer.last)
            ]
            with m.If(ready):
                m.d.sync += buffer.valid.eq(0)

        m.d.sync += ready.eq(1)

        # Fill the intermediate buffer and clear the ready flag.
        with m.If(~buffer.valid):
            m.d.sync += [
                buffer.payload      .eq(self.input.payload),
                buffer.valid        .eq(self.input.valid),
                buffer.last         .eq(self.input.last),
                ready               .eq(0),
            ]
            m.d.comb += self.input.ready.eq(1)
```

For other clock domain conversions we might want a higher counter value, but this class is explicitly for `sync` to `usb` only.